scheduler/src/cook/mesos/util.clj

```diff
@@ -552,3 +552,28 @@
   [ch ch-size]
   (->> (repeatedly ch-size #(async/poll! ch))
        (remove nil?)))
+
+(defn xform-pipe
+  "Reads elements from the `in-chan` channel and supplies elements to the `out-chan` channel.
+   Maintains a state (initialized to `initial-state`) that is updated by applying the reducing function
+   `in-xform` to the current state and the incoming element `(in-xform state element)`.
+   `out-xform` accepts the current state (i.e. `(out-xform state)`) and produces `[reset-state out-value]`.
+   When the `out-chan` channel can receive an element, the `value` is put on the channel and the state is
+   set to the `reset-state` returned from the `out-xform`. Invocation of `out-xform` does not guarantee that
+   `out-value` will be consumed, hence it should not have side-effects. The optional parameter `on-consumed`
+   can be used to register notifications of when a value has been supplied to the `out-chan` channel.
+
+   Note: This function does not perform error handling, exceptions must be explicitly handled in the
+   provided functions (i.e. in-xform, out-xform, on-consumed and on-finished)."
+  [in-chan initial-state in-xform out-chan out-xform &
+   {:keys [on-consumed on-finished] :or {on-consumed identity, on-finished #()}}]
+  (async/go-loop [state initial-state]
+    (let [[reset-state out-value] (out-xform state)
```

Comment:
I think it is ok as is. Though maybe put a note that it is less than ideal. 

If we wanted to go a little crazy.. One thing we could do is have an intermediate-chan with no buffer that we try to write to. That intermediate-chan is connected to the out-chan using one of the pipeline functions (not sure if we want async vs blocking, probably async) and the function between them is to take from the promise-chan. I *think* this would work, but is definitely worth testing.